keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1282 @@
+probes though). This prevents the Pod from being included K8s ready endpoints.
+However this problem will be discovered right away during the testing stage and
+changes will need to be made to sidecar implementation.
+
+Switching to the new sidecars approach will slow down Pod start for Istio. Istio
+today is not blocking other containers to start during its initialization. With
+the switch to the new model, the separation of Initialization stage and main
+containers running stage will be more explicit and many implementations will
+likely wait for sidecar initialization, effectively slowing down Pod startup
+comparing to current approach. This can be eliminated by slight redesign of a
+sidecars.
+
+
+## Design Details
+
+### Backward compatibility
+
+The new field means that any Pod that is not using this field will behave the
+same way as before.
+
+The new field will only work with the proper control plane and kubelet. Upgrade
+and downgrade scenarios will be covered in the further sections.
+
+Outside of Kubernetes-controlled code, there might be third party controllers or
+existing containers relying on the current behavior of Init containers.
+Behaviors they can rely on:
+
+- Assuming nothing is running in the Pod when Init container starts. For
+  instance taking PID of the process assuming there are no sidecars.
+- Incorrectly calculating the resource usage of a Pod not taking the new type of
+  containers into account (e.g. some grafana dashboards may need modification)
+- Stripping the restartPolicy:Always from Init container as an unknown field
+  rendering Pod unfunctional as it will not pass the Init stage after this
+- OPA rules may reject the Pod with new unknown fields because of failure to
+  parse the new field
+
+These potential incompatibilities will be documented.
+
+### Resources calculation for scheduling and pod admission
+
+When calculating whether Pod will fit the Node, resource limits and requests are
+being calculated.
+
+Resources calculation will change for Pod with sidecar containers. Today
+resources are calculated as a maximum between the maximum use of an
+InitContainer and Sum of all regular containers:
+
+`Max ( Max(initContainers), Sum(Containers) )`
+
+With the sidecar containers the formula will change.
+
+Easiest formula would be to assume they are running the duration of the init
+stage as well as regular containers.
+
+`Max ( Max(nonSidecarInitContainers) + Sum(Sidecar Containers), Sum(Sidecar
+Containers) + Sum(Containers) )`
+
+However the true calculations will be different as all Init containers that
+completed before the first sidecar containers will not need to account for any
+sidecar containers for the maximum value calculation.
+
+So the formula, assuming the function:
+
+```
+InitContainerUse(i) = Sum(sidecar containers with index < i) + InitContainer(i)
+```
+
+Is this:
+
+```
+Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(Containers) ) 
```

Comment:
I think [Pod overhead ](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-overhead/) should also be added to all of these calculations, not a big deal in this doc but another easy thing for re-implementors of this calculation to get wrong.

Reply:
added, thanks